aibolit/metrics/npath/main.py

```diff
@@ -104,57 +104,65 @@ def _node_npath(self, node: ASTNode) -> int:
             return self._sequence_npath(node)
 
         def default_handler(node: ASTNode) -> int:
-            return self._sequence_npath(node.children)
+            return self._composite_npath(node)
 
         dispatcher = {
             ASTNodeType.IF_STATEMENT: self._if_npath,
             ASTNodeType.SWITCH_STATEMENT: self._switch_npath,
             ASTNodeType.FOR_STATEMENT: self._for_loop_npath,
             ASTNodeType.WHILE_STATEMENT: self._while_loop_npath,
             ASTNodeType.BINARY_OPERATION: self._binary_expression_npath,
+            ASTNodeType.EXPRESSION: self._expression_npath,
         }
 
         handler = dispatcher.get(node.node_type, default_handler)
         return handler(node)
 
+    def _composite_npath(self, node: ASTNode) -> int:
+        return self._sequence_npath(node.children)
+
     def _sequence_npath(self, nodes: Iterable[ASTNode]) -> int:
-        return math.prod((self._node_npath(child) for child in nodes))
+        return max(1, math.prod(self._node_npath(child) for child in nodes))
 
     def _if_npath(self, node: ASTNode) -> int:
-        condition_npath = self._node_npath(node.condition)
+        condition_npath = self._expression_npath(node.condition)
         then_npath = self._node_npath(node.then_statement)
         else_npath = (
             self._node_npath(node.else_statement)
             if node.else_statement is not None
             else 1
         )
-        return condition_npath * then_npath + else_npath
+        return condition_npath + then_npath + else_npath
 
     def _switch_npath(self, node: ASTNode) -> int:
-        def _group_npath(case_group: ASTNode) -> int:
-            return math.prod((self._node_npath(case) for case in case_group.children))
-
-        case_paths = sum(_group_npath(case_group) for case_group in node.cases)
-        return self._node_npath(node.expression) * max(case_paths, 1)
+        npath = self._expression_npath(node.expression)
+        has_case = False
+        has_default = False
+        for case_group in node.cases:
+            if len(case_group.case):
+                has_case = True
+            else:
+                has_default = True
+            null_statements = max(0, len(case_group.case) - 1)
+            npath += self._node_npath(case_group.statements) + null_statements
+        return npath + int(not has_case) + int(not has_default)
 
     def _for_loop_npath(self, node: ASTNode) -> int:
-        condition_npath = self._condition_npath(node.control)
+        control_npath = sum(self._expression_npath(expr) for expr in node.control.children)
```

Comment:
Can the `self._expression_npath` be reused here?

```suggestion
        control_npath = self._expression_npath(node.control)
```

Reply:
Interesting! Done